### ISSUE TITLE
Cookies should store token, not hashed password

### ISSUE THREAD

**Issue by joshuaadickerson** [elkarte/Elkarte#2344]
It might seem like a small difference at first, but it's huge. Here's how it works:

The user registers. Their password is stored in the database as a hash. A token is generated from a random string and also stored in the database. No separate salt because password_hash() stores it in the hash along with the algorithm.

The user goes to login. They input their username and password. We pull down their password from the database. This is the only point that we are actually going to touch their password. So registration and login are the only times that their password gets pulled down. This means we can add a _huge_ cost to it and even if the database gets compromised, it will be a bitch for the hacker to access it.

The user was authenticated. At the same time we were pulling down their password, we grabbed their authentication token. We create a cookie with that token. We can either serialize the token and user id (using json_encode() and NOT serialize() or just splitting them with some character breaks like \n or a series of characters) or we can create two cookies. Either way, doesn't matter. The point is that there is a cookie with a token and not a hash.

Now that they are logged in, the script redirects to a page. This page has to authenticate the user so it checks their cookie. It does a lookup based on the user this time and checks their token, not their password. The password is never again touched. The password isn't stored on their computer and it isn't transported over the connection.

Even better is that you can logout other users! Instead of having their tokens stored in the member table, they would be stored in an authentication table. That table would contain a bunch of information about the connection. I could say "logout all of my connections from North Korea" or the script can say "you have an active session opened in North Korea. We think this might be Kim Jong-un trying to get the Konami code and recommend you closing that session and changing your password."

**Comment by Spuds** [comment]
I thought the login cookie contained a sha of the password hash plus salt, not just the "raw" password hash.   

Anyway I like the concept of the logout, need to think about this a bit.

**Comment by joshuaadickerson** [comment]
Yeah, it is salted. In this scenario we never send it back to the user. The token is just a random string and we can make that random string as long and as complex as we want. It could be 5 characters. If we don't let them guess anymore than 5 times before we lock their account, it's pretty much impossible for a hacker to guess it. I am not suggesting 5 characters, but it no longer has anything to do with the password.

Currently it looks like this: `list ($id_member, $password, $login_span) = @unserialize($_SESSION['login_' . $cookiename]);`

I would change that to `list($id_member, $token) = json_decode($_SESSION['login_token']);` or similar

This would also be changed:
`	$check = hash('sha256', ($user_settings['passwd'] . $user_settings['password_salt'])) == $password;`

Now it would just do a check of the tokens table. Find the member's id and their token and see if they match. No hashing.

**Comment by joshuaadickerson** [comment]
Another part of that is to silently remove old cookies. Maybe that would be a plugin.

**Comment by joshuaadickerson** [comment]
The hard part for me to figure out is how best to handle the upgrade process without leaving tons of artifacts.

I'm thinking with a table that would be planned to be deleted at some point: old_passwords. 3 fields: id_member, pass, salt. When a person updates their hash, it would remove their row.

After we create that table, the admin then has the option of how long to keep it around for. To ensure that there are no artifacts from that, it would be installed as a plugin. After that information is deleted, the users have to use the forgot password feature.

An option for the admin on upgrade/import would be to force everyone to update their passwords at the time of upgrade/conversion.

**Comment by joshuaadickerson** [comment]
I created a gist to show an example of what I'm talking about and how you get a lot more features out of it https://gist.github.com/joshuaadickerson/142f76c344a998b70961

**Comment by Spuds** [comment]
Cool ... You can probably use TokenHash.class.php for those random strings that you want.
